cupy/cuda/memory.pyx

```diff
@@ -100,11 +101,91 @@ cdef class Memory(BaseMemory):
             self.ptr = runtime.malloc(size)
 
     def __dealloc__(self):
+        # Note: Cannot raise in the destructor! (cython/cython#1613)
         if self.ptr:
-            syncdetect._declare_synchronize()
             runtime.free(self.ptr)
 
 
+cdef inline void async_alloc_check() except*:
+    cdef int dev_id
+    cdef list support = [runtime.deviceGetAttribute(
+        runtime.cudaDevAttrMemoryPoolsSupported, dev_id)
+        for dev_id in range(runtime.getDeviceCount())]
+    _thread_local.device_support_async_alloc = support
+
+
+cdef inline void is_async_alloc_supported(int device_id) except*:
+    if CUDA_VERSION < 11020:
+        raise RuntimeError("memory_async is supported since CUDA 11.2")
+    if runtime._is_hip_environment:
+        raise RuntimeError('HIP does not support memory_async')
+    global is_async_alloc_support_checked
+    if not is_async_alloc_support_checked:
```

Comment:
why not cache this per device, and just do the checks once.
We don't need to check Cuda version or HIP every memory alloc.

Reply:
For the CUDA version and HIP checks, note they are compile-time constants. If evaluated to `True`, Cython will just raise right away, without even generating the rest of code. If evaluated to `False`, those checks are not done at runtime at all.

For the per-device support, could you be more clear? It's already cached in `_thread_local`, but I scratched my head and couldn't think of a way to handle it if this is attempted on a non-supporting device 😅 Don't we have to raise every time one tries it?

Reply:
Yeah, sorry, I was overcomplicating things. Let's leave it as it is. Since the runtime value is cached, we might get right of the global var as well.

Reply:
The global var is used to ensure the runtime API call is done only once (with the cache built). 

Having said that, I think essentially we're comparing the timing of accessing a Python object (cache) with that of making CUDA API calls. To be honest I don't know which one is cheaper 😅 and what's done here is mostly to follow the convention in the codebase, with an assumption that checks like this are generally cheap enough to ignore.